src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -97,11 +104,47 @@ private abstract class AbstractTableDataSourceFindUsagesContext :
                 Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 0);
 
                 // And add ourselves as the source of results for the window.
-                _findReferencesWindow.Manager.AddSource(this);
+                // Additionally, add custom columns to display custom reference information.
+                _findReferencesWindow.Manager.AddSource(this, customColumns.SelectAsArray(c => c.Name));
 
                 // After adding us as the source, the manager should immediately call into us to
                 // tell us what the data sink is.
                 Debug.Assert(_tableDataSink != null);
+
+                // Initialize and update custom column states.
+                // Note that the TableControl.SetColumnStates is not dependent on order of the input column states.
+                _customColumnTitleToStatesMap = GetInitialCustomColumnStates(findReferencesWindow.TableControl.ColumnStates, customColumns);
+                TableControl.SetColumnStates(_customColumnTitleToStatesMap.Values);
+            }
+
+            private static Dictionary<string, ColumnState2> GetInitialCustomColumnStates(
+                IReadOnlyList<ColumnState> allColumnStates,
+                ImmutableArray<AbstractFindUsagesCustomColumnDefinition> customColumns)
+            {
+                var customColumnStatesMap = new Dictionary<string, ColumnState2>(customColumns.Length);
+                var customColumnNames = new HashSet<string>(customColumns.Select(c => c.Name));
+                
+                // Set the default visibility for each custom column.
+                // If there is an existing column state for the custom column, flip it to be non-visible by default at the start of FAR query.
+                // We do so because the column will have empty values for all results for a FAR query for certain cases such as types, literals, no references found case, etc.
+                // It is preferable to dynamically hide an empty column for such queries, and dynamically show the column if it has at least one non-default value.
```

Comment:
tiny suggestion. i really like the "rewrap" extension.  It helps so you can define some sort of good length for comments (i do 100 col), enabling you to easily edit comments, and just hit ctrl-q to reflow it.  That way you don't end up either having these really wide comments, or having to go through the pain of manually reflowing things.  Just a suggestion :)

Reply:
it's probably worth mentioning what/where actually causes the columns to go into the visible state.

Also, FAR windows are reused.  What happens between calls.  Do things flip back to non-visible.  Or once visible does it just stay that way?

Reply:
Improved the comments at the `TableControl.SetColumnStates` callsite above.

---
In reply to: [222057639](https://github.com/dotnet/roslyn/pull/30155#discussion_r222057639) [](ancestors = 222057639)